net: fix race condition in self-connect detection

theStack opened the pull request:
This PR fixes a recently discovered race condition in the self-connect detection (see #30362 and #30368).

Initiating an outbound network connection currently involves the following steps after the socket connection is established (see [`CConnman::OpenNetworkConnection`](https://github.com/bitcoin/bitcoin/blob/bd5d1688b4311e21c0e0ff89a3ae02ef7d0543b8/src/net.cpp#L2923-L2930) method):
1. set up node state
2. queue VERSION message (both steps 1 and 2 happen in [`InitializeNode`](https://github.com/bitcoin/bitcoin/blob/bd5d1688b4311e21c0e0ff89a3ae02ef7d0543b8/src/net_processing.cpp#L1662-L1683))
3. add new node to vector `m_nodes`
    
If we connect to ourself, it can happen that the sent VERSION message (step 2) is received and processed locally *before* the node object is added to the connection manager's `m_nodes` vector (step 3). In this case, the self-connect remains undiscovered, as the detection doesn't find the outbound peer in `m_nodes` yet (see `CConnman::CheckIncomingNonce`).
    
Fix this by swapping the order of 2. and 3., by taking the `PushNodeVersion` call out of `InitializeNode` and doing that in the `SendMessages` method instead, which is only called for `CNode` instances in `m_nodes`.

The temporarily reverted test introduced in #30362 is readded. Fixes #30368.

Thanks go to vasild, mzumsande and dergoegge for suggestions on how to fix this (see https://github.com/bitcoin/bitcoin/issues/30368#issuecomment-2200625017 ff. and https://github.com/bitcoin/bitcoin/pull/30394#discussion_r1668290789).

theStack committed:
net: fix race condition in self-connect detection

Initiating an outbound network connection currently involves the
following steps after the socket connection is established (see
 `CConnman::OpenNetworkConnection` method):
    1. set up node state
    2. queue VERSION message
    3. add new node to vector `m_nodes`

If we connect to ourself, it can happen that the sent VERSION message
(step 2) is received and processed locally *before* the node object
is added to the connection manager's `m_nodes` vector (step 3). In this
case, the self-connect remains undiscovered, as the detection doesn't
find the outbound peer in `m_nodes` yet (see `CConnman::CheckIncomingNonce`).

Fix this by swapping the order of 2. and 3., by taking the `PushNodeVersion`
call out of `InitializeNode` and doing that in the `SendMessages` method
instead, which is only called for `CNode` instances in `m_nodes`.

Thanks go to vasild, mzumsande, dergoegge and sipa for suggestions on
how to fix this.

theStack committed:
Reapply "test: p2p: check that connecting to ourself leads to disconnect"

This reverts commit 9ec2c53701a391629b55aeb2804e8060d2c453a4 with
a tiny change included (identation of the wait_until call).

tdb3 reviewed (commented):
Approach ACK

Great job finding the bug and fixing it.

To sanity check, re-arranged calls to mimick failure recreation (https://github.com/bitcoin/bitcoin/issues/30368#issuecomment-2200034205).  `p2p_handshake` failed as expected after the re-arrange.

```diff
diff --git a/src/net.cpp b/src/net.cpp
index a10be3e33a..6ed167d6fd 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2921,6 +2921,8 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
     pnode->grantOutbound = std::move(grant_outbound);
 
     m_msgproc->InitializeNode(*pnode, nLocalServices);
+    m_msgproc->SendInitialMessages(*pnode);
+    std::this_thread::sleep_for(std::chrono::seconds{1});
     {
         LOCK(m_nodes_mutex);
         m_nodes.push_back(pnode);
@@ -2928,7 +2930,7 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
         // update connection count by network
         if (pnode->IsManualOrFullOutboundConn()) ++m_network_conn_counts[pnode->addr.GetNetwork()];
     }
-    m_msgproc->SendInitialMessages(*pnode);
+    //m_msgproc->SendInitialMessages(*pnode);
 }
```

Also started on signet, connected to peers, and synced the latest ~1500 blocks successfully.

The name `SendInitialMessages()` seems reasonable to me, since it's generic enough to adapt to future change without rename.

I'd like to double check @vasild's comment (https://github.com/bitcoin/bitcoin/issues/30368#issuecomment-2202728919) before full ACK.

>In case you settle with adding to m_nodes before PushNodeVersion() and I do not manage to review, it is good to ensure that no code will be upset by an existence of an (oubound) entry in m_nodes that does not have version pushed.

dergoegge commented on src/net.h:
```diff
@@ -991,9 +991,12 @@ class NetEventsInterface
     virtual void InitializeNode(CNode& node, ServiceFlags our_services) = 0;
 
+    /** Queue initial messages for peer (only relevant for outbound connections) */
+    virtual void SendInitialMessages(CNode& node) = 0;
```
I think I'd be nicer to not change this interface and I believe it would be trivial to just use `SendMessages`? i.e. push the version message to outbound peers at the start of `SendMessages` (if it hasn't been pushed already).

dergoegge replied:
Please put that state on `Peer` if you go this route.

dergoegge commented on src/net.h:
```diff
@@ -991,7 +991,7 @@ class NetEventsInterface
     static Mutex g_msgproc_mutex;
 
-    /** Initialize a peer (setup state, queue any initial messages) */
+    /** Initialize a peer (setup state) */
```
(not in this PR but) Since we don't want any messages send in `InitializeNode`, maybe it makes sense to disallow that somehow, perhaps by not passing a `CNode`?

dergoegge commented on src/net_processing.cpp:
```diff
@@ -5817,6 +5817,12 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
     if (MaybeDiscourageAndDisconnect(*pto, *peer)) return true;
 
+    // Initiate version handshake for outbound connections
+    if (!pto->IsInboundConn() && !peer->m_initial_version_message_sent) {
+        PushNodeVersion(*pto, *peer);
+        peer->m_initial_version_message_sent = true;
+    }
```
This isn't fully atomic, i.e. in theory two threads could enter this section at the same time.

Not tested but maybe the following works:
```suggestion
    if (!pto->IsInboundConn() && peer->m_initial_version_message_sent.compare_exchange_strong(false, true)) {
        PushNodeVersion(*pto, *peer);
    }
```

Alternatively, we could guard `m_initial_version_message_sent` with `NetEventsInterface::g_msgproc_mutex`, since it is only accessed from the "msghand" thread.

theStack committed:
net: prevent sending messages in `NetEventsInterface::InitializeNode`

Now that the queueing of the VERSION messages has been moved out of
`InitializeNode`, there is no need to pass a mutable `CNode` reference any
more. With a const reference, trying to send messages in this method would
lead to a compile-time error, e.g.:

----------------------------------------------------------------------------------------------------------------------------------
...
net_processing.cpp: In member function ‘virtual void {anonymous}::PeerManagerImpl::InitializeNode(const CNode&, ServiceFlags)’:
net_processing.cpp:1683:21: error: binding reference of type ‘CNode&’ to ‘const CNode’ discards qualifiers
 1683 |     PushNodeVersion(node, *peer);
...
----------------------------------------------------------------------------------------------------------------------------------

dergoegge reviewed (approved):
Code review ACK fb587ce36afcc8789214af45a36082c4f5238e50

maflcko reviewed (commented):
left two nits.

furszy reviewed (commented):
Code review ACK fb587ce36afcc8789214af45a36082c4f5238e50

Nice finding.

glozow reviewed (commented):
lgtm, do you plan to take any of the suggestions?

mzumsande reviewed (commented):
I wonder if this could this lead to problems with "non-shy" software that would send VERSION immediately on incoming connections (which bitcoin core doesn't do thanks to Hal Finney's #101, but alternative clients might).

When Bitcoin Core makes an outgoing connection after this patch, it won't send out the version message immediately, so it could receive a version message before sending one. In that case, I think that `ProcessMessage` could send out other messages (WTXIDRELAY etc.) before sending the version message, which would be a protocol violation.

One possible fix for this might be to not process messages of outbound peers before having sent the version message.

I didn't test any of this, this is just from reading the code.

dergoegge commented:
Option 2. seems the best to me.

furszy commented:
> When Bitcoin Core makes an outgoing connection after this patch, it won't send out the version message immediately, so it could receive a version message before sending one. In that case, I think that ProcessMessage could send out other messages (WTXIDRELAY etc.) before sending the version message, which would be a protocol violation.

Can we get a test for this? Fixes are great but tests are awesome to not break it again in the future.

mzumsande commented:
> Can we get a test for this? Fixes are great but tests are awesome to not break it again in the future.

Our own test framework was behaving that way before #29353, so maybe the functional tests would have caught it? I thought about reverting c340503b67585b631909584f1acaa327f5af25d3 locally to find out, but didn't have the time yet.

mzumsande reviewed (commented):
ACK 16bd283b3ad05daa41259a062aee0fc05b463fa6

> Our own test framework was behaving that way before #29353, so maybe the functional tests would have caught it? I thought about reverting [c340503](https://github.com/bitcoin/bitcoin/commit/c340503b67585b631909584f1acaa327f5af25d3) locally to find out, but didn't have the time yet.

FYI, I tested reverting c340503b67585b631909584f1acaa327f5af25d3 on top of this branch now and without the latest fix (2) to ProcessMessages, other messages are actually sent before `version`, and `p2p_add_connection.py` fails intermittently - so the fix works.

tdb3 reviewed (approved):
ACK 16bd283b3ad05daa41259a062aee0fc05b463fa6

glozow reviewed (commented):
ACK 16bd283b3ad05daa41259a062aee0fc05b463fa6

dergoegge reviewed (approved):
ACK 16bd283b3ad05daa41259a062aee0fc05b463fa6

mzumsande commented:
I didn't find a reference in a BIP (there isn't any for the general version message flow as far as I can see) and maybe "violation" is too strong, but any message received before the version would be ignored by bitcoin core peers:
https://github.com/bitcoin/bitcoin/blob/46878326808f643f08ec3f69dcec5c8fafeb7b59/src/net_processing.cpp#L3889-L3893
So  this is clearly not the intended way the protocol should work.

fanquake commented:
Backported to 27.x in #30467.
